src/compiler/visitorPublic.ts

```diff
@@ -22,35 +22,61 @@ import {
 /**
  * Visits a Node using the supplied visitor, possibly returning a new Node in its place.
  *
+ * - If the input node is undefined, then the output is undefined.
+ * - If the visitor returns undefined, then the output is undefined.
+ * - If the output node is not undefined, then it will satisfy the test function.
+ * - In order to obtain a return type that is more specific than `Node`, a test
+ *   function _must_ be provided, and that function must be a type predicate.
+ *
  * @param node The Node to visit.
  * @param visitor The callback used to visit the Node.
  * @param test A callback to execute to verify the Node is valid.
  * @param lift An optional callback to execute to lift a NodeArray into a valid Node.
  */
-export function visitNode<T extends Node>(node: T, visitor: Visitor | undefined, test?: (node: Node) => boolean, lift?: (node: readonly Node[]) => T): T;
-
+export function visitNode<TIn extends Node | undefined, TVisited extends Node | undefined, TOut extends Node>(
+    node: TIn,
+    visitor: Visitor<NonNullable<TIn>, TVisited> | undefined,
+    test: (node: Node) => node is TOut,
+    lift?: (node: readonly Node[]) => Node,
+): TOut | (TIn & undefined) | (TVisited & undefined);
 /**
  * Visits a Node using the supplied visitor, possibly returning a new Node in its place.
  *
+ * - If the input node is undefined, then the output is undefined.
+ * - If the visitor returns undefined, then the output is undefined.
+ * - If the output node is not undefined, then it will satisfy the test function.
+ * - In order to obtain a return type that is more specific than `Node`, a test
+ *   function _must_ be provided, and that function must be a type predicate.
+ *
  * @param node The Node to visit.
  * @param visitor The callback used to visit the Node.
  * @param test A callback to execute to verify the Node is valid.
  * @param lift An optional callback to execute to lift a NodeArray into a valid Node.
  */
-export function visitNode<T extends Node>(node: T | undefined, visitor: Visitor | undefined, test?: (node: Node) => boolean, lift?: (node: readonly Node[]) => T): T | undefined;
-
-export function visitNode<T extends Node>(node: T | undefined, visitor: Visitor | undefined, test?: (node: Node) => boolean, lift?: (node: readonly Node[]) => T): T | undefined {
-    if (node === undefined || visitor === undefined) {
+export function visitNode<TIn extends Node | undefined, TVisited extends Node | undefined>(
+    node: TIn,
+    visitor: Visitor<NonNullable<TIn>, TVisited> | undefined,
+    test?: (node: Node) => boolean,
+    lift?: (node: readonly Node[]) => Node,
+): Node | (TIn & undefined) | (TVisited & undefined);
+export function visitNode(
+    node: Node,
+    visitor: Visitor | undefined,
+    test?: (node: Node) => boolean,

```

Comment:
I'm sort of worried about doing that, because it's an actual behavior change (which this PR mostly avoids); if we actually require people pass in the test function, then any existing code which uses this API will no longer work if they weren't passing in the test before. We could make it still behave like now and return the node without testing, but that seems scary.

That, and I think that there may be places where we don't care about checking the type because the node that you get back doesn't matter. Maybe there aren't so many places like that out there? I can test this at least on our codebase quick.

Reply:
When I wrote this years ago, the original intent was that external consumers should always pass a `test`, but internally we could opt not to except in a debug build. Somehow that line got murkier over time, however. If 5.0 is an opportunity to make breaking API changes, I'm strongly inclined to make this one of them.

Reply:
I tried making this change; there were a few cases where it wasn't clear what kind of thing was to be returned (sometimes it wasn't used at all, or were `return visitNode(node, visitor)`), then the 3-4 cases where we use visitNode to do node cloning (and more about the inability to pass a non-type-predicate test downward).

I'm personally still of the opinion that it's worth allowing a fallback that's just `Node` for cases where the return type doesn't matter or it isn't used. Maybe we could make that overload internal, but it does seem useful in general. Not sure how to evaluate that for existing codebases, though.